Initialize parser pool

gbehrmann opened the pull request:
Motivation:

It appears that OSGSAMLBootstrap#bootstrap is supposed to replace
DefaultBootstrap#bootstrap, but in contrast to the latter the former fails to
initialize the parser pool. I suspect the parser pool initialization was added
during an upgrade of opensaml library and wasn't mirrored in the xacml client
library.

Modification:

Add a call to initializeParserPool during bootstrapping.

Result:

No null pointer exception when using the XACML client.

gbehrmann committed:
Initialize parser pool

Motivation:

It appears that OSGSAMLBootstrap#bootstrap is supposed to replace
DefaultBootstrap#bootstrap, but in contrast to the latter the former fails to
initialize the parser pool. I suspect the parser pool initialization was added
during an upgrade of opensaml library and wasn't mirrored in the xacml client
library.

Modification:

Add a call to initializeParserPool during bootstrapping.

Result:

No null pointer exception when using the XACML client.

bbockelm commented:
LGTM!  Thanks!

bbockelm commented:
By the way - what version of opensaml are you targeting?   The default one in the pom or something else?

gbehrmann commented:
That was the default one in the pom (2.4.1). 

Just a little context so you know what I am up: I recently ported dCache from JGlobus to CANL, but in the last minute I discovered that the XACML client depends on JGlobus. I have a patch ready that ports the XAML client to the latest version of the voms lib and to CANL and thus can get rid of JGlobus. If you are interested in that I will submit a pull request (I still need to test it).

The reason why dCache moves away from JGlobus is that JGlobus has appeared more or less unmaintained for a long while and it is a huge dependency for dCache. Due to bouncy castle breaking the API often, JGlobus is now forcing us to use old versions of other libraries as JGlobus depends specifically on bouncy castle 1.46 (eg we could not upgrade to newer versions of the sshd lib due to this).

bbockelm commented:
Sure, I'd be happy to see the patch.

The JGlobus dependency for OSG is also mostly centered around SRM (well, the ill-fated HTTPG transport layer and the GridFTP java client).  How do you avoid this?

In terms of this particular client, I think the jglobus dependencies are pretty shallow: mostly around the client certificate handling.  I certainly hope CANL has this well-covered!

gbehrmann commented:
Yes, CANL handles all the trust store and cert chain validation nicely and with a much more convenient API.

As for GSI and the FTP client for dCache:
- I reimplemented GSI without the GSSAPI dependency. GSI is just TLS with a delegation step after the initial handshake (and an optional framing around the TLS frames). That's not difficult to implement. I chose to make it complicated by implementing it as a decorator for SSLEngine, mostly so I could hook this into the non-blocking I/O stuff in Jetty 9.  You can have a look at https://github.com/dCache/dcache/tree/master/modules/common-security/src/main/java/org/dcache/gsi 
- The Axis 1 support I replaced too by implementing support for the Apache HTTP components client in Axis and then adding a simple delegation step on top of that, see https://github.com/dCache/dcache/blob/master/modules/srm-common/src/main/java/org/dcache/srm/client/HttpClientSender.java, https://github.com/dCache/dcache/blob/master/modules/srm-common/src/main/java/org/dcache/srm/client/GsiHttpClientSender.java, https://github.com/dCache/dcache/blob/master/modules/srm-common/src/main/java/org/dcache/srm/client/GsiConnectionSocketFactory.java.
- As for the FTP client, I copied the entire thing from JGlobus and updated it to use CANL and dCache classes were appropriate. Over time I will see if I can unify the data channel handling with what we use server side (server side FTP has always been pure dCache). See https://github.com/dCache/dcache/blob/master/modules/ftp-client/src/main/java/org/dcache/ftp/client/extended/GridFTPControlChannel.java for the implementation of GSI in the client. Since FTP is inherently GSSAPI like, I am relying on my own simplified version of that which I call DSSAPI :-) That's not intended to be generic in any way and doesn't come with the complications of full GSSAPI. The plugability of GSSAPI was never utilized anyway.

Another benefit of getting rid of JGlobus is that we get strict RFC 2818 support (and in dCache it isn't optional). That's still going to be a problem for some sites at the moment, but by the time the next golden release gets out (next summer), this should have been solved.

Once all this settles down, I will talk to the CANL developers and ask whether they are interested in some of the GSI stuff.

bbockelm commented:
Ah yes - I forgot that bestman had already forked off the Jetty work you had done.  I'll keep an eye on your upstream for when we start thinking about Jetty 9.

In case you want to look at the state of bestman2, we are working on a cleanup of the source repo (upstream worked in a private CVS repo, and each release was a single commit to SVN ... among other weirdness) here: https://github.com/edquist/bestman2

Forking off the FTP client is probably a step too far for us: I'd rather spend our (limited) energy on this stack on porting people to other clients.

gbehrmann commented:
@bbockelm, did you get my mail about having a look at the patch on my fork? I would appreciate if you could have a look - its okay if you don't have time; just say so. Thanks.
